core/src/main/scala/tech/beshu/ror/accesscontrol/factory/decoders/rules/GroupsRuleDecoder.scala

```diff
@@ -45,7 +59,7 @@ class GroupsRuleDecoder(usersDefinitions: Definitions[UserDef],
       .emapE { groups =>
         NonEmptyList.fromList(usersDefinitions.items) match {
           case Some(userDefs) =>
-            Right(RuleWithVariableUsageDefinition.create(new GroupsRule(GroupsRule.Settings(groups, userDefs), caseMappingEquality)))
+            Right(RuleWithVariableUsageDefinition.create(new GroupsRule(AbstractGroupsRule.Settings(groups, userDefs), caseMappingEquality)))
```

Comment:
we create here GroupsRule, but it should depend on the context of a child-class. Maybe we can add a creator to the constructor. Like this:

```
private abstract class BaseGroupsRuleDecoder[R <: BaseGroupRule](creator: BaseGroupRule.Settings => R,
 usersDefinitions: Definitions[UserDef],
 implicit val caseMappingEquality: UserIdCaseMappingEquality)
  extends RuleBaseDecoderWithoutAssociatedFields[R] { ... }
  ```

and use it here to create a rule instance.

Reply:
Oh yes, this won't work at all. In fact it throws syntax error for "groups" rule. I tried to generalise it as suggested, but I got ultra stuck with implicits and type parameters. Can you help? 🥺